pkg/activator/podip_watcher.go

```diff
@@ -0,0 +1,258 @@
+package activator
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging/logkey"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/apis/serving"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/network/prober"
+	"knative.dev/serving/pkg/queue"
+	"knative.dev/serving/pkg/reconciler"
+	"knative.dev/serving/pkg/resources"
+)
+
+const probePortName string = "http"
+
+type revisionWatcher struct {
+	dests           []string
+	healthyDests    map[string]bool
+	healthyDestsMux sync.RWMutex
+
+	transport http.RoundTripper
+	destsChan <-chan []string
+	logger    *zap.SugaredLogger
+}
+
+func NewRevisionWatcher(destsChan <-chan []string, transport http.RoundTripper,
+	logger *zap.SugaredLogger) *revisionWatcher {
+	rw := revisionWatcher{
+		healthyDests: make(map[string]bool),
+		transport:    transport,
+		destsChan:    destsChan,
+		logger:       logger,
+	}
+	return &rw
+}
+
+// getHealthyDests returns a list of l4 dests which are "healthy"
+// (have returned a success on the most recent probe)
+func (rw *revisionWatcher) getHealthyDests() []string {
+	rw.healthyDestsMux.RLock()
+	defer rw.healthyDestsMux.RUnlock()
+
+	ret := []string{}
```

Comment:
presize to the `healthyDests` size. Also move creation outside of the lock.

Reply:
+1 on static size, I dont think I can move creation outside of lock if I am relying on healthyDests size though

Reply:
actually, I cant static size because we need to filter based on whether each dest is healthy